test/e2e/lib/creation.go

```diff
@@ -0,0 +1,117 @@
+	messagingv1alpha1 "github.com/google/knative-gcp/pkg/apis/messaging/v1alpha1"
+	pubsubv1alpha1 "github.com/google/knative-gcp/pkg/apis/pubsub/v1alpha1"
+)
+
+func (c *Client) CreateUnstructuredObjOrFail(spec *unstructured.Unstructured) {
+	namespace := c.Namespace
+	gvr, _ := meta.UnsafeGuessKindToResource(spec.GroupVersionKind())
+	_, err := c.Core.Dynamic.Resource(gvr).Namespace(namespace).Create(spec, v1.CreateOptions{})
+	if err != nil {
+		c.T.Fatalf("Failed to create object %q: %v", spec.GetName(), err)
+	}
+	c.Tracker.Add(gvr.Group, gvr.Version, gvr.Resource, namespace, spec.GetName())
+}
+
+func (c *Client) CreateChannelOrFail(channel *messagingv1alpha1.Channel) {
+	channels := c.KnativeGCP.MessagingV1alpha1().Channels(c.Namespace)
+	_, err := channels.Create(channel)
+	if err != nil {
+		c.T.Fatalf("Failed to create channel %q: %v", channel.Name, err)
+	}
+	c.Tracker.AddObj(channel)
+}
+
+func (c *Client) CreateAuditLogsOrFail(auditlogs *eventsv1alpha1.AuditLogsSource) {
+	auditlogses := c.KnativeGCP.EventsV1alpha1().AuditLogsSources(c.Namespace)
+	_, err := auditlogses.Create(auditlogs)
+	if err != nil {
+		c.T.Fatalf("Failed to create auditlogs %q: %v", auditlogs.Name, err)
+	}
+	c.Tracker.AddObj(auditlogs)
+}
+
+func (c *Client) CreatePubSubOrFail(pubsub *eventsv1alpha1.PubSub) {
+	pubsubs := c.KnativeGCP.EventsV1alpha1().PubSubs(c.Namespace)
+	_, err := pubsubs.Create(pubsub)
+	if err != nil {
+		c.T.Fatalf("Failed to create pubsub %q: %v", pubsub.Name, err)
+	}
+	c.Tracker.AddObj(pubsub)
+}
+
+func (c *Client) CreateStorageOrFail(storage *eventsv1alpha1.Storage) {
+	storages := c.KnativeGCP.EventsV1alpha1().Storages(c.Namespace)
+	_, err := storages.Create(storage)
+	if err != nil {
+		c.T.Fatalf("Failed to create storage %q: %v", storage.Name, err)
+	}
+	c.Tracker.AddObj(storage)
+}
+
+func (c *Client) CreatePullSubscriptionOrFail(pullsubscription *pubsubv1alpha1.PullSubscription) {
+	pullsubscriptions := c.KnativeGCP.PubsubV1alpha1().PullSubscriptions(c.Namespace)
+	_, err := pullsubscriptions.Create(pullsubscription)
+	if err != nil {
+		c.T.Fatalf("Failed to create pullsubscription %q: %v", pullsubscription.Name, err)
+	}
+	c.Tracker.AddObj(pullsubscription)
+}
+
+// WithServiceForJob returns an option that creates a Service binded with the given job.
+func WithServiceForJob(name string) func(*batchv1.Job, *Client) error {
+	return func(job *batchv1.Job, client *Client) error {
+		namespace := job.Namespace
+		svc := resources.ServiceDefaultHTTP(name, job.Spec.Template.Labels)
+
+		svcs := client.Core.Kube.Kube.CoreV1().Services(namespace)
+		if _, err := svcs.Create(svc); err != nil {
+			return err
+		}
+		client.Tracker.Add("", "v1", "services", namespace, name)
+		return nil
+	}
+}
+
+func (c *Client) CreateJobOrFail(job *batchv1.Job, options ...func(*batchv1.Job, *Client) error) {
+	// set namespace for the job in case it's empty
+	namespace := c.Namespace
+	job.Namespace = namespace
```

Comment:
but if it's set, you are overriding it? Maybe:

`if job.Namespace == "" {
    job.Namespace = c.Namespace
}`

Reply:
Done. Thanks.